### PR TITLE
Validate Ksonnet apps through component dir presence

### LINKED ISSUE #594: Better detection of ksonnet apps
Some false positives when scanning https://github.com/coreos/prometheus-operator.git for apps: 

image

### PR DESCRIPTION
**Pull request by merenbach**
Closes #594

### CODE REVIEW

**Review by jessesuen — DISMISSED** [review]
Based on the code changes, this is not addressing the issue I filed in the bug. The problem is that we are blindly assuming any directory with an app.yaml is a ksonnet app dir, and including it in the app list API.

**Comment by merenbach** [comment]
@jessesuen mea culpa! Great catch. I believe this is now addressed with the latest commit, which causes the Argo CD UI to only show the Helm charts, and no Ksonnet apps, in the example repo provided in the original bug.

File: server/repository/repository.go [hunk 1]
```diff
@@ -100,8 +100,11 @@ func (s *Server) ListApps(ctx context.Context, q *RepoAppsQuery) (*RepoAppsRespo
 
 	items := make([]*AppInfo, 0)
 
-	for i := range ksonnetRes.Items {
-		items = append(items, &AppInfo{Type: string(repository.AppSourceKsonnet), Path: ksonnetRes.Items[i]})
+	for _, i := range ksonnetRes.Items {
+		t, err := repository.IdentifyAppSourceTypeByAppPath(i)
```
Comment (jessesuen) [inline server/repository/repository.go]: Making this call per app.yaml is *way* too expensive for the app list query. We need a different mechanism to identify ksonnet apps by file lists alone.

Comment (merenbach) [reply server/repository/repository.go]: @jessesuen have pushed a change to address this

File: server/repository/repository.go [hunk 2]
```diff
@@ -100,8 +100,11 @@ func (s *Server) ListApps(ctx context.Context, q *RepoAppsQuery) (*RepoAppsRespo
 
 	items := make([]*AppInfo, 0)
 
-	for i := range ksonnetRes.Items {
-		items = append(items, &AppInfo{Type: string(repository.AppSourceKsonnet), Path: ksonnetRes.Items[i]})
+	for _, i := range ksonnetRes.Items {
+		t := repository.IdentifyAppSourceTypeByAppDir(filepath.Dir(i))
```
Comment (jessesuen) [inline server/repository/repository.go]: Does this code even work? Based on the implementation of IdentifyAppSourceTypeByAppDir (), you need direct access to the git repo on the local filesystem, which the API server does not.

Comment (merenbach) [reply server/repository/repository.go]: @jessesuen okay, I think we're a little better now. :D

File: server/repository/repository.go [hunk 3]
```diff
@@ -100,8 +104,13 @@ func (s *Server) ListApps(ctx context.Context, q *RepoAppsQuery) (*RepoAppsRespo
 
 	items := make([]*AppInfo, 0)
 
-	for i := range ksonnetRes.Items {
-		items = append(items, &AppInfo{Type: string(repository.AppSourceKsonnet), Path: ksonnetRes.Items[i]})
+	for _, k := range ksonnetRes.Items {
+		for _, c := range componentRes.Items {
```
Comment (jessesuen) [inline server/repository/repository.go]: Can we avoid a n^2 algorithm for identifying ksonnet apps? Maybe iterate ksonnet apps first, store them as candidates in a map, then validate that they have adjacent component directories while you iterate componentRes.

Comment (merenbach) [reply server/repository/repository.go]: O(2n) algorithm in place now.

**Review by alexmt — APPROVED** [review]

**Comment by jessesuen** [comment]
@merenbach this change broke listing of ksonnet apps completely. Was this tested?

image

**Comment by merenbach** [comment]
> @merenbach this change broke listing of ksonnet apps completely. Was this tested?
> 
> image

@jessesuen have issued PR #730 to revert this merge. I tested this, but less effectively than I needed to, is what it looks like.

**Pull request merged**